src/dbnode/persist/fs/commitlog/commit_log.go

```diff
@@ -506,69 +538,154 @@ func (l *commitLog) write() {
-		l.metrics.flushErrors.Inc(1)
-		l.log.Error("failed to flush commit log", zap.Error(err))
+		if err != nil {
+			l.metrics.errors.Inc(1)
+			l.metrics.flushErrors.Inc(1)
+			l.log.Error("failed to flush commit log", zap.Error(err))
 
-		if l.commitLogFailFn != nil {
-			l.commitLogFailFn(err)
+			if l.commitLogFailFn != nil {
+				l.commitLogFailFn(err)
+			}
 		}
-	}
 
-	// onFlush only ever called by "write()" and "openWriter" or
-	// before "write()" begins on "Open()" and there are no other
-	// accessors of "pendingFlushFns" so it is safe to read and mutate
-	// without a lock here
-	if len(l.pendingFlushFns) == 0 {
-		l.metrics.flushDone.Inc(1)
-		return
-	}
+		// onFlush will never be called concurrently. The flushFn for the primaryWriter
+		// will only ever be called synchronously by the single-threaded writer goroutine
+		// and the flushFn for the secondaryWriter will only be called by the asynchronous
+		// goroutine (created by the single-threaded writer) when it calls Close() on the
+		// secondary (previously primary due to a hot-swap) writer during the reset.
+		//
+		// Note that both the primary and secondar's flushFn may be called during calls to
+		// Open() on the commitlog, but this takes place before the single-threaded writer
+		// is spawned which precludes it from occurring concurrently with either of the
+		// scenarios described above.
+		if len(writer.pendingFlushFns) == 0 {
+			l.metrics.flushDone.Inc(1)
+			return
+		}
 
-	for i := range l.pendingFlushFns {
-		l.pendingFlushFns[i](callbackResult{
-			eventType: flushEventType,
-			err:       err,
-		})
-		l.pendingFlushFns[i] = nil
+		for i := range writer.pendingFlushFns {
+			writer.pendingFlushFns[i](callbackResult{
+				eventType: flushEventType,
+				err:       err,
+			})
+			writer.pendingFlushFns[i] = nil
+		}
+		writer.pendingFlushFns = writer.pendingFlushFns[:0]
+		l.metrics.flushDone.Inc(1)
 	}
-	l.pendingFlushFns = l.pendingFlushFns[:0]
-	l.metrics.flushDone.Inc(1)
 }
 
 // writerState lock must be held for the duration of this function call.
-func (l *commitLog) openWriter() (persist.CommitLogFile, error) {
-	if l.writerState.writer != nil {
-		if err := l.writerState.writer.Close(); err != nil {
-			l.metrics.closeErrors.Inc(1)
-			l.log.Error("failed to close commit log", zap.Error(err))
-
-			// If we failed to close then create a new commit log writer
-			l.writerState.writer = nil
+func (l *commitLog) openWriters() (persist.CommitLogFiles, error) {
+	// Ensure that the previous asynchronous reset of the secondary writer (if any)
+	// has completed before attempting to start a new one and/or modify the writerState
+	// in any way.
+	l.waitForSecondaryWriterAsyncResetComplete()
+
+	if l.writerState.primaryWriter.writer == nil || l.writerState.secondaryWriter.writer == nil {
+		// If either of the commitlog writers is nil then open both of them synchronously. Under
+		// normal circumstances this will only occur when the commitlog is first opened.
```

Comment:
Is there a valid scenario where one writer is nil while the other isn't? If not, should you check for it? Do we need to close the writer if on of them isn't nil?

Reply:
There is a valid scenario and you're right about the closing if one or the other is not already nil. I'll update it